1-static-analysis.md

```diff
@@ -0,0 +1,298 @@
+# Surface Module Source Static Analysis
+
+## Synopsis
+
+Extend instances of `ModuleSource` such that they reflect certain results of
+static analysis, like their `import` and `export` bindings, such that tools can
+inspect module graphs.
+
+## Dependencies
+
+This proposal depends on [Module and ModuleSource][0] from the [Compartments
+proposal](README.md) to introduce `ModuleSource`.
+
+## Design
+
+Extend [ModuleSource][0], such that instances have the following properties:
+
+- `bindings`, an `Array` of `Binding`s.
+- `needsImportMeta`, a `boolean` indicating that the module contains
```

Comment:
I don't think this is meta that we should expose, and if we decide to expose, how reliable is this? e.g.: what happen if the module source does `eval("import.meta")`, what does this do today?

Reply:
agreed - adding or removing use of `import.meta` should never become observable, and thus a breaking change.

Reply:
Presumably `needsImportMeta` of a module source is the product of observing the most recent source. This is intended to communicate to `importHook` whether it needs to construct a heavy `import.meta` (including a `resolve` closure) or not, and is meaningful to the Node.js community.

Reply:
That's not an optimization that should be baked into the spec, imo.

Reply:
I don't think the potential of that optimization is worth starting to expose a previously secret implementation detail.

Reply:
It is not a very good secret. Anyone in a position to call `new ModuleSource(source)` is in an equally good position to obtain the same information from Babel at extraordinary cost.

Reply:
Sure - that's if you're able to obtain the source, which in most environments isn't a guarantee.

Either way, I think that's something a userland hook should be forced to do if it wants to depend on implementation details, just like user code has to do to peer inside a function's closure.

Reply:
To be clear, I do not particularly care about this detail, but I do care about this API being a viable alternative to Node.js’s `Loader`, which has this optimization because it makes a considerable difference to cold start times in a module graph with 1,000 modules where 5 of them use `import.meta.resolve` and the others don’t mention `import.meta` at all. Without this hint, they need a considerably number of `resolve` closures that will never be observed.

Reply:
Is there any reason they couldn't be getters on `import.meta` that lazily create those closures?

Reply:
getter is a function too. They also have costs to create.

Reply:
First, you are right that there are cases where a `ModuleSource` will be available but the source text will not. That becomes possible with either a harmonious import reflection or deferred execution proposal, where a CSP-vetted `ModuleSource` would be exposed but not necessarily its source text. In my opinion, that strengthens the case for exposing static analysis, but I read your disagreement.

Currently, hosts can only provide objects with a null prototype. We could allow `new Module()` to receive an arbitrary value to present as `import.meta`. If that were the case, such an object could have a host-defined prototype with accessors, provided that any of the computed properties of one `import.meta` could be inferred from other properties of `import.meta`. To wit, an `import.meta.resolve` that closes over `import.meta.url`, `import.meta.path`, or some other state accessible to the implementation of the `Meta` class (pun intended).

I read in your preference not to expose static analysis as flowing from a respectable abundance of caution. I have no objection to leaving this question unanswered until some time has been allowed to uncover a genuine hazard that we currently avoid by not revealing the static analysis of successfully-linked CSP-vetted modules, and in parallel giving some time to surface objections to this alternative approach of allowing arbitrary `import.meta` values.

Reply:
> That's not an optimization that should be baked into the spec, imo.

So to be clear, the way `import.meta` works already allows hosts to lazily create `import.meta` [when it's accessed](https://tc39.es/ecma262/#sec-meta-properties-runtime-semantics-evaluation). This proposal is even more conservative than that in that rather than waiting until access-time, it only tells when the whole module is to be executed.

Specifically this behaviour using Node's experimental vm modules is fully spec compliant today:

```js
const module = new SourceTextModule(`
   console.log(1);
   void import.meta; // access import.meta
   console.log(3);
`, {
    initializeImportMeta(meta) {
        console.log(2);
    },
});
```

and prints `1,2,3`.

This proposal is more conservative in that you can't observe when `import.meta` is first accessed, just when the module is executed (so you could only print `2` before or after execution of the module in that example).